draft-ietf-tls-esni.md

```diff
@@ -446,33 +490,114 @@ to harmonize these to make sure that we maintain key separation.]]
-with {{!RFC7540}}; Section 9.2.)
-
-If the server does not negotiate TLS 1.3 or above, then the client
-MUST abort the connection with an "unsupported_version" alert.
-If the server does not provide an "encrypted_server_name" extension
-in EncryptedExtensions, the client MUST abort the connection with
-an "illegal_parameter" alert. Moreover, it MUST check that
-ClientESNIInner.nonce matches the value of the
-"encrypted_server_name" extension provided by the server,
-and otherwise abort the connection with an "illegal_parameter"
-alert.
-
-## Client-Facing Server Behavior
+The client MUST place the value of ESNIKeys.public_name in the "server_name"
+extension. (This is required for technical conformance with {{!RFC7540}};
+Section 9.2.)
+
+If the server negotiates TLS 1.3 or above and provides an
+"encrypted_server_name" extension in EncryptedExtensions, the client
+then processes the extension's "response_type" field:
+
+- If the value is "esni_accept", the client MUST check that the extension's
+  "nonce" field matches ClientESNIInner.nonce and otherwise abort the
+  connection with an "illegal_parameter" alert. The client then proceeds
+  with the connection as usual, verifying the certificate against the desired
+  name.
+
+- If the value is "esni_retry_request", the client proceeds with the handshake,
+  verifying the certificate against ESNIKeys.public_name as described in
+  {{verify-public-name}}. If the handshake completes successfully, the client
+  MUST abort the connection with an "esni_required" alert.
+
+  The client then processes the "retry_keys" field from the server's
+  "encrypted_server_name" extension. If one of the values used a version known
+  to the client, the client SHOULD retry the handshake with a new transport
+  connection, using that value to encrypt the SNI. If no value is applicable,
+  the client SHOULD retry with ESNI disabled.
+
+  The client MUST NOT use the server-provided retry keys until the handshake
+  completes successfully. On success, it MUST NOT overwrite the DNS-provided
+  keys with the retry keys. It MUST use the retry keys at most once and
+  continue offering DNS-provided keys for subsequent connections. This avoids
+  introducing a tracking vector, should a malicious server present
+  client-specific retry keys.
+
+- If the field contains any other value, the client MUST abort the connection
+  with an "illegal_parameter" alert.
+
+If the server negotiates an earlier version of TLS, or if it does not
+provide an "encrypted_server_name" extension in EncryptedExtensions, the
+client proceeds with the handshake, verifying the certificate against
+ESNIKeys.public_name as described in {{verify-public-name}}. The client MUST
+NOT enable the False Start optimization {{RFC7918}}. If the handshake completes
+successfully, the client MUST abort the connection with an "esni_required" alert.
+It then SHOULD retry the handshake with a new transport connection and ESNI
+disabled.
```

Comment:
The SHOULD here is either out of scope, or under-specified.

I think that the intent here is to say that if ESNI is not negotiated for any reason, but the connection is completed successfully with "public_name", then you aren't being attacked, but ESNI has simply been disabled.  A necessary condition for the "SHOULD retry" here is that the connection is successfully established with "public_name", but that connection is not explicit here.

You also want to add a condition here that if the server provides retry keys that the client cannot use, then it can regard ESNI as securely disabled.

If that is not the intent, then this stipulates policy: it's OK to fall back when you have a clear signal that ESNI is enabled, but the server doesn't do ESNI.  That's not a policy I would support.